core/src/main/java/de/danoeh/antennapod/core/service/playback/PlaybackService.java

```diff
@@ -464,37 +457,32 @@ public int onStartCommand(Intent intent, int flags, int startId) {
         final boolean castDisconnect = intent.getBooleanExtra(EXTRA_CAST_DISCONNECT, false);
         Playable playable = intent.getParcelableExtra(EXTRA_PLAYABLE);
         if (keycode == -1 && playable == null && !castDisconnect) {
-            Log.e(TAG, "PlaybackService was started with no arguments");
-            stopService();
+            // Typical cases when the service was started with no argument
+            // - when it is first bound, and then moved to startedState, as in <code>serviceManager.moveServiceToStartedState()</code>
+            // - callers (e.g., Controller) explicitly
+            Log.d(TAG, "PlaybackService was started with no arguments.");
             return Service.START_NOT_STICKY;
         }
 
-        if ((flags & Service.START_FLAG_REDELIVERY) != 0) {
-            Log.d(TAG, "onStartCommand is a redelivered intent, calling stopForeground now.");
-            stopForeground(true);
-        } else {
-            if (keycode != -1) {
-                Log.d(TAG, "Received media button event");
-                boolean handled = handleKeycode(keycode, true);
-                if (!handled) {
-                    stopService();
-                    return Service.START_NOT_STICKY;
-                }
-            } else if (!flavorHelper.castDisconnect(castDisconnect) && playable != null) {
-                started = true;
-                boolean stream = intent.getBooleanExtra(EXTRA_SHOULD_STREAM,
-                        true);
-                boolean startWhenPrepared = intent.getBooleanExtra(EXTRA_START_WHEN_PREPARED, false);
-                boolean prepareImmediately = intent.getBooleanExtra(EXTRA_PREPARE_IMMEDIATELY, false);
-                sendNotificationBroadcast(NOTIFICATION_TYPE_RELOAD, 0);
-                //If the user asks to play External Media, the casting session, if on, should end.
-                flavorHelper.castDisconnect(playable instanceof ExternalMedia);
-                if (playable instanceof FeedMedia) {
-                    playable = DBReader.getFeedMedia(((FeedMedia) playable).getId());
-                }
-                mediaPlayer.playMediaObject(playable, stream, startWhenPrepared, prepareImmediately);
+        if (keycode != -1) {
+            Log.d(TAG, "Received media button event");
+            boolean handled = handleKeycode(keycode, true);
+            if (!handled) {
+                serviceManager.stopService(); // TODO: [2716]-RemoveOptional why is it added in the first place in v1.7.0?
```

Comment:
What does this mean?

Reply:
@ByteHamster  This is something I have meant to ask - please see https://github.com/AntennaPod/AntennaPod/pull/2954#issuecomment-451778876

Reply:
This was added in an attempt to fix the phantom notification. If AntennaPod is not playing and some unknown media button event is routed to AntennaPod, the service will be started but no notification will be shown (or the phantom notification will not be replaced).

> if someone connects a new headset with some controls that AntennaPod does not recognize, the ongoing playback will stop.

You are right. If it is currently playing, it would be better to not kill the service.

Reply:
removed the call in the latest commits

Reply:
I think just removing the call might lead to `Service did not call startForeground` errors. It should only be ignored if the service is currently playing

Reply:
You're right - the `stopService()` should be replace with a `servicecManager.stopIfInBackground()` .

Reply:
@ByteHamster the issue turns out to be even more trickier. I'd like your feedback before checking in changes.

Android 8+'s foreground service requirement is more stringent: when a service is started as foreground, e.g., via `ContextCompat.startForegroundService()`, 
- the service **MUST** bring up a foreground notification, i.e., `service.startForeground(int notificationId, ...)`
- if the service stops itself without calling `service.startForeground()`, Android will still kill the app:

```
2019-03-06 13:43:33.451 E/ActivityManager: ANR in de.danoeh.antennapod.debug
    PID: 19977
    Reason: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{772dfe0 u0 de.danoeh.antennapod.debug/de.danoeh.antennapod.core.service.playback.PlaybackService}
```

---
**Applying to the case of unhandled key code:**
There are three sub-cases:
1. the service is playing some media: we call `service.startForeground()` with another notification to satisfy Android's requirement.
2. the service playback is paused (but the service is still running in background): we can also call `service.startForeground()`. However, the side effect is that the service is made foreground. The notification cannot be swiped away.
3. the service has not been running (before the keycode): we can't call the normal `service.startForeground()` as there is no media to construct it.
The app will still be killed.

To address the issues
- Approach 1: 
  - case 2: we call `service.startForeground()` to make the service foreground again, then bring it back as background again.
  - case 3: make the service foreground with a separate dummy notification, and then stop the service. 
- Approach 2: for cases 2 and 3, simply stop the service (or do nothing), and let Android kill the app. Given nothing is playing, the actual interruption to users are minimal.

What's your thought?


---
Note: testing unhandled key code:

I did some manual testing by hacking [the widget code](https://github.com/AntennaPod/AntennaPod/blob/d26d2126072654ba95f31f2bb638479bc2f199c4/core/src/main/java/de/danoeh/antennapod/core/service/PlayerWidgetJobService.java#L145) to generate an unsupported button event, e.g., `KEYCODE_MEDIA_TOP_MENU`

Reply:
That's bad. So it is basically impossible to fully fix the phantom notification because we still need to show a dummy notification... :/

> Approach 2: for cases 2 and 3, simply stop the service (or do nothing), and let Android kill the app. Given nothing is playing, the actual interruption to users are minimal.

I don't like this approach. It pops up a crash notification, users can report the crash and Google will display the crash in the developer console. I don't really like approach 1 either, but it's less bad :P

Just for reference, you have probably read that already, here is Google telling that it is intended behavior https://issuetracker.google.com/issues/76112072

Reply:
@ByteHamster Thanks for pointing out Google's official stance.

I believe there is a way to fix the issue altogether, but it requires a larger refactoring on `PlaybackService` handles media buttons in general.

The premise: instead of passing media button events via `startService`, pass the events `binding` to the service and invokes relevant methods.

Investigation that backed the premise
- I used support library's standard [v4.meida.session.MediaButtonReceiver](https://developer.android.com/reference/android/support/v4-media-session/MediaButtonReceiver), instead of AntennaPod's own to handle it.
- using it does not cause the service not in foreground exception.
- while it is not fully functional either (the media button events are not processed by `PlaybackService` correctly, as if they are all unsupported), the logcat shows that PlaybackService was connected (rather than started), by the receiver to try to do something, and it does not cause service not in foreground exception.

```
# MediaButtonReceiver2 is a receiver that extends support library's 
2019-03-07 10:25:29.225 D/MediaButtonReceiver2: onReceive() called with: action = [android.intent.action.MEDIA_BUTTON], extras = [Bundle[mParcelledData.dataSize=176]], 
2019-03-07 10:25:29.242 D/PlaybackService: Service created.
2019-03-07 10:25:29.258 D/MediaSessionService: Media button session is changed to de.danoeh.antennapod.debug/PlaybackService (userId=0)
2019-03-07 10:25:29.273 D/PlaybackService: Received onBind event
2019-03-07 10:25:29.275 D/PlaybackService: OnGetRoot: clientPackageName=de.danoeh.antennapod.debug; clientUid=10071 ; rootHints=Bundle[{}]
2019-03-07 10:25:29.280 D/PlaybackService: Received onUnbind event
2019-03-07 10:25:29.280 D/PlaybackService: Service is about to be destroyed
```

---
**Options to address the issue**
1. Approach 1 above (i.e., create a dummy notification if necessary)
while creating yet another dummy notification is quite distasteful, at least this one's scope is well defined (with a clear termination point)

2. Approach 2 above (i.e., let Android kill the app when there is nothing playing)
While it is not ideal, it is still an improvement over the status quo in v1.7.1.

3. Change AntennaPod's [`MediaButtonReceiver`](https://github.com/AntennaPod/AntennaPod/blob/1.7.1/core/src/main/java/de/danoeh/antennapod/core/receiver/MediaButtonReceiver.java) so that it passes the media button event by binding to `PlaybackService` instead of starting the service, mimic how support library does it.
I have not tried it yet, but based on the investigation so far it should work.

4. Replace  AntennaPod's `MediaButtonReceiver` with support library
It has the potential of simplifying `PlaybackService` by pushing the work to support library, e.g., most of `handleKeyCode()`. However, it is a larger piece of work:
4a. requires larger refactoring: as shown above, `PlaybackService` as-is does not play well with support library's MediaButtonReceiver yet (which uses `MediaBrowserServiceCompat` to interact with the service, implying there is part of support's `MediaBrowserServiceCompat` that `PlaybackService` has not properly implemented).
4b. it is unclear yet how we can supply AntennaPod-specific media button logic, e.g., custom handling for `MEDIA_{STOP_NEXT,PREVIOUS}`

Thoughts?

Reply:
> Replace AntennaPod's MediaButtonReceiver with support library
It has the potential of simplifying PlaybackService by pushing the work to support library

Well, if you have the time and motivation to do this, I would support that :) Simplifying hat huge and buggy PlaybackService class is definitely a good idea. I fear that doing so will delay this PR. That would be sad because I see it as a blocker for the next release (mainly because of the rxjava stuff). I want to merge this rather sooner than later. Maybe using approach 1 (dummy notification) for now and then cleaning up the service later in another PR would be the way to go.

Reply:
I took a stab with approach 3 (passing media button event by binding to `PlaybackService` rather than starting it), and want to get some feedback (the changes are not merged in this PR yet): 

https://github.com/orionlee/AntennaPod/commit/75e93ddfee4866ac39aff4ea7ebc1bed03790330

In terms of lines of codes, it is probably longer than dummy notification / approach 1. However, the codes are quite straightforward, while dummy notification / approach 1  will require adding more special cases in the already-complicated `setupNotification()`.
 
If we settle on the approach, I will do some more clean-up / refactoring and merge the changes to the PR.

Reply:
That looks a bit dirty. Especially since Google explicitly states the following:

> https://developer.android.com/guide/components/bound-services.html#Binding
> Note: Only activities, services, and content providers can bind to a service—you can't bind to a service from a broadcast receiver.

Apparently, doing so can also make the app crash:

> https://stackoverflow.com/q/18931670/4193263
> ReceiverCallNotAllowedException: BroadcastReceiver components are not allowed to bind to services

Reply:
1. While google doc says one can't bind to a service in BroadcastReceiver, I believe it is safe in this case. The support lib's MediaButtonReceiver does indeed bind to the service ([source](https://android.googlesource.com/platform/frameworks/support/+/refs/tags/android-9.0.0_r34/media/src/main/java/androidx/media/session/MediaButtonReceiver.java#118)), indirectly via `MediaBrowserCompat.connect()` (which ultimately does the service binding in platform's [`MedirBrowserCompat.connect()`](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/media/java/android/media/browse/MediaBrowser.java#187) .

2. `ReceiverCallNotAllowedException` is avoided in the code by using application's context rather than BroadcastReceiver's. (Support lib does the same thing)

3. Support lib's MediaButtonReceiver does one more thing: it calls [BoradcastReceiver.goAsync()](https://developer.android.com/reference/android/content/BroadcastReceiver.html#goAsync()) to give itself a bit more leeway. It does not use heavier approaches, e.g., pushing the logic to a `JobIntentService`.

---
I think the code is safe enough, as long as we add the `goAsync()` calls. Thoughts?

Reply:
Here is a revised one with `goAsync()`: https://github.com/orionlee/AntennaPod/commit/8bb720b374ded40db65a546db06e4a61ca515a09

I stress-tested it, by pressing play (on a widget) on a paused streaming podcast while network is down, which causes the play ultimately fail after ~30 seconds. There is no app crash / ANR , on Android 9 and Android 4.4 emulators.

```
2019-03-10 13:06:19.998 D/MediaButtonReceiver: Received intent
2019-03-10 13:06:19.999 V/MediaButtonReceiver: onReceive() returned.
2019-03-10 13:06:20.000 V/MediaButtonReceiver: in Handler.post() start
2019-03-10 13:06:20.003 V/MediaButtonReceiver: in Handler.post() end
2019-03-10 13:06:20.007 D/PlaybackService: Service created.
2019-03-10 13:06:20.036 D/PlaybackService: Received onBind event
2019-03-10 13:06:20.040 V/MediaButtonReceiver: onServiceConnected() - starts.  thread:main
2019-03-10 13:06:20.040 D/PlaybackService: Handling keycode: 85
2019-03-10 13:06:20.052 V/MediaButtonReceiver: onServiceConnected() - about to unbind and pendingResult.finish()
2019-03-10 13:06:20.055 D/PlaybackService: OnStartCommand called
2019-03-10 13:06:20.071 D/PlaybackService: Received onUnbind event
2019-03-10 13:06:20.078 V/PlaybackService: onPlaybackStateChange(8)
...
2019-03-10 13:06:20.125 V/PlaybackService: notificationSetupTask: playerStatus=PREPARING
2019-03-10 13:06:20.132 V/PlaybackService: notificationSetupTask: make service foreground
...
# timed out after about 30 seconds
2019-03-10 13:06:53.435 V/PlaybackService: onPlaybackStateChange(2)
2019-03-10 13:06:53.437 D/PlaybackService: notificationSetupTask: Starting background work
2019-03-10 13:06:53.439 V/PlaybackService: notificationSetupTask: playerStatus=PAUSED
# There is no crash or ANR
```

Reply:
@ByteHamster Let me know what you think regarding unsupported key button handling.